internal/gatewayapi/listener.go

```diff
@@ -172,7 +175,182 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR resource
+					hostname2: string(ptr.Deref(httpsListeners[j].Hostname, "")),
+				}
+				overlappingListeners[j] = &overlappingListener{
+					listener1: string(httpsListeners[j].Name),
+					listener2: string(httpsListeners[i].Name),
+					hostname1: string(ptr.Deref(httpsListeners[j].Hostname, "")),
+					hostname2: string(ptr.Deref(httpsListeners[i].Hostname, "")),
+				}
+			}
+		}
+	}
+
+	for i, listener := range httpsListeners {
+		if overlappingListeners[i] != nil {
+			status.SetGatewayListenerStatusCondition(gateway.Gateway,
+				listener.listenerStatusIdx,
+				gwapiv1.ListenerConditionOverlappingTLSConfig,
+				metav1.ConditionTrue,
+				gwapiv1.ListenerReasonOverlappingHostnames,
+				fmt.Sprintf(
+					"The hostname %s overlaps with the hostname %s in listener %s",
+					overlappingListeners[i].hostname1,
+					overlappingListeners[i].hostname2,
+					overlappingListeners[i].listener2,
+				),
+			)
+			if listener.httpIR != nil {
+				listener.httpIR.TLSOverlaps = true
+			}
+		}
+	}
+}
+
+func checkOverlappingCertificates(gateway *GatewayContext) {
+	httpsListeners := []*ListenerContext{}
+	for _, listener := range gateway.listeners {
+		if listener.Protocol == gwapiv1.HTTPSProtocolType {
+			httpsListeners = append(httpsListeners, listener)
+		}
+	}
+
+	type overlappingListener struct {
+		listener1 string
+		listener2 string
+		san1      string
+		san2      string
 	}
+
+	overlappingListeners := make([]*overlappingListener, len(httpsListeners))
+	for i := range httpsListeners {
+		if overlappingListeners[i] != nil {
+			continue
+		}
+		for j := i + 1; j < len(httpsListeners); j++ {
+			if overlappingListeners[j] != nil {
+				continue
+			}
+			if httpsListeners[i].Port != httpsListeners[j].Port {
+				continue
+			}
+
+			overlappingCertificate := isOverlappingCertificate(httpsListeners[i].certDNSNames, httpsListeners[j].certDNSNames)
+			if overlappingCertificate != nil {
+				// Overlapping listeners can be more than two, we only report the first two for simplicity.
+				overlappingListeners[i] = &overlappingListener{
+					listener1: string(httpsListeners[i].Name),
+					listener2: string(httpsListeners[j].Name),
+					san1:      overlappingCertificate.san1,
+					san2:      overlappingCertificate.san2,
+				}
+				overlappingListeners[j] = &overlappingListener{
+					listener1: string(httpsListeners[j].Name),
+					listener2: string(httpsListeners[i].Name),
+					san1:      overlappingCertificate.san2,
+					san2:      overlappingCertificate.san1,
+				}
+			}
+		}
+	}
+
+	for i, listener := range httpsListeners {
+		if overlappingListeners[i] != nil {
+			status.SetGatewayListenerStatusCondition(gateway.Gateway,
+				listener.listenerStatusIdx,
+				gwapiv1.ListenerConditionOverlappingTLSConfig,
+				metav1.ConditionTrue,
+				gwapiv1.ListenerReasonOverlappingCertificates,
+				fmt.Sprintf(
+					"The certificate san %s overlaps with the certificate san %s in listener %s",
```

Comment:
can we add a line saying ALPN has been set to http/1.1 to avoid connection coalescing issues seen on h2 ?